### ISSUE TITLE
How to run Handlers example?

### ISSUE THREAD

**Issue by sairus7** [OxygenFramework/Oxygen.jl#197]
I'm trying to use example code for different handlers, and there is something wrong with stream and websocket handlers. I don't see new text appearing in browser with `/stream` query, and a Websocket handshake failed with `/ws` query.
What am I doing wrong here?

Server side:
```julia
using HTTP
using Oxygen

# Request Handler
@get "/" function(req::HTTP.Request)
    return "hello"
end

# Stream Handler
@stream "/stream" function(stream::HTTP.Stream)
    startwrite(stream)
    write(stream, "stream started\n")
    for i in 1:10
        write(stream, "$i\n")
        sleep(1.0)
    end
    write(stream, "stream ended")
    closewrite(stream)
end

# Websocket Handler
@websocket "/ws" function(ws::HTTP.WebSocket)
    for msg in ws
        # simple echo server
        send(ws, msg)
    end
end

# start the web server
serve(async = true)
```

Client side:
```julia
using HTTP, Sockets

# host = Sockets.getipaddr()
# host = IPv4("0.0.0.0")
host = ip"127.0.0.1"
port = 8080

resp = HTTP.request("GET", "http://$host:$port/")
@info resp # return null after /stream query

# there is no data chunks shown in console
resp = HTTP.open(:GET, "http://$host:$port/stream") do http
    @info http 
    data = read(http, String)
    @info data
end

# error: Websocket handshake failed
WebSockets.open("ws://$host:$port/ws") do ws
    write(ws, "hi websockets")
    for msg in ws
        @info msg
        # send(ws, resp)
    end
end
```

Also, running root `/` after `/stream` somehow returns response from previous stream query (and then sometimes returns null instead of "hello").

**Comment by ndortega** [comment]
Hi @sairus7 

Thanks for the code snippet! Here's some tips and general guidelines for setting these routes up

Streaming

- Include headers in the sream
  ```julia
  HTTP.setheader(stream, "Content-Type" => "text/plain")
  HTTP.setheader(stream, "Transfer-Encoding" => "chunked")
  ```

Websockets
- I believe your first line in your websocket route is breaking it. You can't call the write() function on a websocket to send a message

When it comes to connecting a client to these routes I'd recommend using an application like postman to quickly verify that things are working.

For your streaming issue, this looks like a similar thing I ran into when writing tests for streaming. When connecting to streams I had to set the `Connection` header to `"close"` to flush out all the content. More specifically, this turns off persistent connections and forces it to close after each request. 

 You can see more in the [Streaming Tests file](https://github.com/OxygenFramework/Oxygen.jl/blob/master/test/streamingtests.jl). It's worth mentioning I only had this issue when connecting to them via julia and didn't have any issues when calling it through a browser or postman.

```julia
response = HTTP.post("$localhost/api/post/chunked/text",  headers=Dict("Connection" => "close"))
```

**Comment by sairus7** [comment]
Thanks, I've changed my initial example, using `headers=Dict("Connection" => "close")` in `/stream` request fixed the problem.

server:
```julia
using HTTP
using Oxygen

# Request Handler
@get "/hello" function(req::HTTP.Request)
    return "hello"
end

# Stream Handler
@stream "/stream" function(stream::HTTP.Stream)
    HTTP.setheader(stream, "Access-Control-Allow-Origin" => "*")
    HTTP.setheader(stream, "Content-Type" => "text/html; charset=utf-8") # "text/plain"
    HTTP.setheader(stream, "Transfer-Encoding" => "chunked")
    #HTTP.setheader(io, "Connection" => "keep-alive")

    startwrite(stream)
    write(stream, "stream started\n")
    for i in 1:5
        write(stream, "$i\n")
        sleep(1.0)
    end
    write(stream, "stream ended")
    closewrite(stream)
end

# Websocket Handler
@websocket "/ws" function(ws::HTTP.WebSocket)
    try
        for msg in ws
            send(ws, "Received message: $msg")
        end
    catch e
        if !isa(e, HTTP.WebSockets.WebSocketError)
            rethrow(e)
        end
    end
end

# start the web server
serve(async = true)
```

client:
```julia
using HTTP, Sockets
using HTTP.WebSockets
# host = Sockets.getipaddr()
# host = IPv4("0.0.0.0")
host = ip"127.0.0.1"
port = 8080

resp = HTTP.get("http://$host:$port/hello")
@info resp

HTTP.open(:GET, "http://$host:$port/stream") do io # headers=Dict("Connection" => "close")
    # write(io, "Hi streaming")
    # resp = startread(io) # -> HTTP.Response]
    while !eof(io)
        data = readavailable(io)
        text = String(data)
        @info "read: $text"
    end
end #-> HTTP.Response

WebSockets.open("ws://$host:$port/ws") do ws
    send(ws, "Hi websockets")
    response = receive(ws)
    @info response
end
```

However, without `"Connection" => "close"` there are several strange behaviours:
1. Repeatedly running `/stream` returns immediately with no console output,
2. Running `/ws` after `/stream` fails with `"Websocket handshake failed"`
3. Running `/hello` after `/stream` returns null in response body. 
4. Sometimes I event get `/hello` responce somehow leaked into /stream @info

**Comment by ndortega** [comment]
I'm glad that helped fix your issue!

I also ran into most of those issues when running without that header. It's been a couple of months since I've gone digging through the HTTP.jl docs, but I'm pretty sure they try to reuse the same connection object between requests from the same client to help performance.

Unlike the examples in HTTP.jl where you basically create a new server for a websocket, Oxygen uses a single server for requests, streams and websockets and upgrades connection requests on the fly. Setting that header should make it not lean to heavily on its persistence logic.

If I had to guess, there might be some cross over in persistence between requests that's causing that "leaking" you found. I will say that I've had no issues connecting to websockets or streaming when using plain js or postman - so this might just by a HTTP.jl client thing.
